salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+                                                             on_open_callback=self._on_connection_open,
+                                                             on_open_error_callback=self._on_connection_error,
+                                                             custom_ioloop=self._io_loop)
+
+        # set up exchange for message broadcast
+        self._fanout_exchange_name = self.queue_name
+
+    def _on_connection_open(self, connection):
+        """
+        Invoked by pika when connection is opened successfully
+        :param connection:
+        :return:
+        """
+        self._channel = connection.channel(on_open_callback=self._on_channel_open)
+
+    def _on_connection_error(self, connection, exception):
+        """
+        Invoked by pika when connection on connection error
+        :param connection:
+        :param exception:
+        :return:
+        """
+        log.error("Failed to connect", exc_info=True)
+
+    def _on_channel_open(self, channel):
+        """
+        Invoked by pika when channel is opened successfully
+        :param channel:
+        :return:
+        """
+        self._channel.exchange_declare(exchange=self._fanout_exchange_name, exchange_type=ExchangeType.fanout, callback=self.on_exchange_declare)
+
+    def on_exchange_declare(self, method):
+        """Invoked by pika when RabbitMQ has finished the Exchange.Declare RPC
+        command.
+        """
+        log.debug("Exchange declared: {}".format(self._fanout_exchange_name))
+        self._channel.queue_declare(queue=self.queue_name, callback=self._on_queue_declared)
+
+    def _on_queue_declared(self, method):
+        """
+        Invoked by pika when queue is declared successfully
+        :param method:
+        :return:
+        """
+        log.debug("Queue declared: {}".format(self._queue_name))
+
+        self._channel.queue_bind(
+            self.queue_name,
+            self._fanout_exchange_name,
+            routing_key=self.queue_name,
+            callback=self._on_queue_bind)
+
+    def _on_queue_bind(self, method):
+        """
+        Invoked by pika when queue bound successfully
+        :param method:
+        :return:
+        """
+
+        log.debug("Queue bound [{}]".format(self.queue_name))
+
+        def _callback_wrapper(channel, method, properties, payload):
+             if self._callback:
+                log.debug("Calling message callback")
+                return self._callback(payload)
+
+        self._channel.basic_consume(self.queue_name, _callback_wrapper, auto_ack=True)
+
+    def register_message_callback(self, callback):
+        """
+        Register a callback that receives message on a queue
+        :param callback:
+        :return:
+        """
+        self._callback = callback
+
+
+class AsyncRabbitMQReqChannel(salt.transport.client.ReqChannel):
+    """
+    Encapsulate sending routines to RabbitMQ.
+
+    RMQ Channels default to 'crypt=aes'
```

Comment:
**note** I have no clue if it comes into play for RMQ, but seeing this brought it to mind - I know that there are certain times that we want FIPS complicance, and we've got some flags for that. If that's a consideration for any type of crypts that RMQ needs to blocklist, then that would be something to account for.